README.md
<img src="https://crashcatch.com/images/logo.png" width="150">

# Introduction
The Flutter Crash Catch Library allows you to monitor crash and error reports across your Flutter projects 
running on Android, iOS, Windows Desktop and Web and report crashes and errors to Crash Catch (https://crashcatch.com). 

# Installing
The first thing to do is  to add the Crash Catch library to your pubspec.yaml file under the dependencies. An example is below, ensure that the version number is the latest version number that is shown on the GitHub repositories tags. 

``` yaml
dependencies:
  flutter:
    sdk: flutter
	crashcatchlib_flutter: ^2.1.1
```

Then run the Pub get command to update the dependencies as below:
```
flutter pub get
```

# Using the Library
If you project has multiple different screens, then you need to pass an instance of Crash Catch to the class Constructor (this is the recommended and most efficient use) or initialise on each new screen, and have a helper method that you can from anywhere to initalise and be able to update your API key or project ID in one place if you ever need to. 

Import the Crash Catch project into your class, as shown below:
``` Dart
import 'package:crashcatchlib_flutter/CrashCatch.dart';
```

Then create an instance of the library and passing in a BuildContext and send the initialisation request as follows:

``` Dart
class Home extends StatelessWidget {
	@override
  Widget build(BuildContext context) {
	CrashCatch crashCatch = new CrashCatch(context);
	crashCatch.initialiseCrashCatch("<project_id>", "<api_key>", "<project_version>");
	return (
		//Create your widget here
	);
  }
}
```
In the above example, you should update <your_project_id> and <api_key> from the project settings page which can be copied directly to your clipboard from the Crash Catch website. The <project_version> should be updated on each version release. 

If all you are worried about is capturing unhandled exceptions, then that's all you need to do, however, you report errors that occur within try/catch blocks as and when required. 

# Reporting Errors
Unhandled crashes and errors are automatically reported once Crash Catch is initialised. However, you can report errors that are caught within a try/catch block. 

This is done by calling the reportCrash method within the CrashCatch object. This takes two parameters, the exception object, and the severity, and an optional third parameter being the stack. 

If you don't want to pass the stack, then that's not a problem, if you don't the Crash Catch library, will automatically detect the stack for the current thread regardless and use that to send as part of the error. The second parameter is the severity, which can be one of three options, LOW, MEDIUM and HIGH using the Severity enum. 

This is done using the following example:

```Dart
try
{
	throw Exception("Something has gone wrong");
}
on Exception catch (e)
{
	crashCatch.reportCrash(e, Severity.LOW);
}
```

Alternatively, you can pass the stack to the reportCrash method as the 3rd parameter. This is done in the example below:

```
try
{
	throw Exception("Something has gone wrong");
}
on Exception catch (e, stacktrace)
{
	crashCatch.reportCrash(e, Severity.LOW, stacktrace);
}
```

There's also a fourth optional parameter on the reportCrash method where you can pass a Map<String, dynamic> object of custom properties, which can be used to provide custom information to help add debug information to help you diagnose why an error occurred. An example of this is blow:

```
try
{
	throw Exception("Something has gone wrong");
}
on Exception catch (e, stacktrace)
{
	crashCatch.reportCrash(e, Severity.HIGH, stack: stacktrace, customProperties: {
    	"key 1": "value 1",
		"key 2": "value 2",
		"Key 3": 3
	});
}
```

That's it you are fully set up to report crashes and errors on your flutter mobile projects. 

CHANGELOG.md
## [1.0.0]
* Initial release

## [1.0.1]
* Fixes issue with Load Balancer cookie
* Fixed issue with checking for blank session id

## [1.0.2]
* Bug fixes from pre-release testing

## [1.0.3]
* Bug fixes for cookie handling to production servers during pre-release testing

## [1.0.4]
* Bug fixes for load balancer cookie handling to production servers during pre-release testing

## [2.0.1]
* Updated dependencies

## [2.1.1]
* Fixed issue where library was treated as legacy as wasn't fully null type safe
* Added support for additional platforms that are in stable release by flutter (Android, iOS, Windows and Web)
* Removed some unnecessary logging

test/crashcatchlib_flutter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:crashcatchlib_flutter/crashcatchlib_flutter.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/CrashCatch.dart
library crashcatchlib_flutter;
import 'dart:collection';
import 'dart:io';
import 'dart:convert';
import 'dart:io' show Platform;

import 'package:flutter/cupertino.dart';
import 'package:flutter/services.dart';
import 'package:http/http.dart' as http;
import 'package:http/http.dart';
import 'package:device_info/device_info.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:math';
import 'dart:convert';

enum Severity {
  LOW,
  MEDIUM,
  HIGH
}

enum CrashType {
  HANDLED,
  UNHANDLED
}

class CrashCatch
{
  String _projectId = "";
  String _versionName = "";
  String _deviceId;
  String _apiKey = "";
  String _sessionId = "";
  String _doLb = "";
  bool _initialisationCompleted = false;
  BuildContext _context;
  static List<HashMap<String, String>> _crashQueue;

  CrashCatch(BuildContext buildContext) {
    this._context = buildContext;
    CrashCatch._crashQueue = [];
  }

  void _setupUnhandledException() {
    FlutterError.onError = (FlutterErrorDetails details) {
      _sendUnhandledCrash(details);
    };
  }

  void initialiseCrashCatch(String projectId, String apiKey, String version) async
  {
    this._projectId = projectId;
    this._apiKey = apiKey;
    this._versionName = version;

    //Check the app shared preferences to see if the device id has been set, if
    //not generate a random device id string
    final SharedPreferences prefs = await SharedPreferences.getInstance();

    this._deviceId = prefs.getString("crashcatch_device_id") ?? "";

    if (this._deviceId == "")
    {
      this._deviceId = generateRandomString();
      await prefs.setString("crashcatch_device_id", this._deviceId);
    }

    HashMap requestData = new HashMap<String, dynamic>();
    requestData["ProjectID"] = projectId;
    requestData["DeviceID"] = _deviceId;
    requestData["AppVersion"] = version;
    _sendRequest("initialise", requestData);

    this._setupUnhandledException();

  }

  void reportCrash(Exception exception, Severity severity, {StackTrace stack, Map<String, dynamic> customProperties = const {} }) async
  {
    String stacktrace = stack == null ? StackTrace.current.toString() : stack
        .toString();

    HashMap<String, String> requestData = await returnPostData(exception, stacktrace, severity, CrashType.HANDLED, customProperties);

    if (_initialisationCompleted) {
      _sendRequest("crash", requestData);
    }
    else
    {
      CrashCatch._crashQueue.add(requestData);
    }
  }

  void _sendUnhandledCrash(FlutterErrorDetails flutterErrorDetails) async
  {
    HashMap<String, String> requestData = await returnPostData(flutterErrorDetails, flutterErrorDetails.stack.toString(), Severity.HIGH, CrashType.UNHANDLED, {});
    if (_initialisationCompleted)
    {
      _sendRequest("crash", requestData);
    }
    else
      {
        CrashCatch._crashQueue.add(requestData);
      }
  }

  Future<HashMap<String, String>> returnPostData(Object exception, String stacktrace, Severity severity, CrashType crashType, Map<String, dynamic> customProperties) async
  {
    HashMap<String, String> decodedStack = _decodeStacktrace(stacktrace, crashType);

    String osVersion = "";
    String apiVersion = "";
    DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
    if (Platform.isAndroid) {
      AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
      osVersion = androidInfo.version.release;
      apiVersion = androidInfo.version.sdkInt.toString();
    }
    else if (Platform.isIOS) {
      IosDeviceInfo iosInfo = await deviceInfo.iosInfo;
      osVersion = "iOS " + iosInfo.systemVersion;
    }

    int width = MediaQuery
        .of(_context)
        .size
        .width
        .ceil();
    int height = MediaQuery
        .of(_context)
        .size
        .height
        .ceil();

    String defaultLocale = Platform.localeName;

    HashMap requestData = new HashMap<String, String>();

    if (exception is Exception)
    {
      Exception exceptionObj = exception;
      requestData["ExceptionType"] = exceptionObj.runtimeType.toString();
      requestData["ExceptionMessage"] = exceptionObj.toString();
    }
    else
    {
      FlutterErrorDetails exceptionObj = exception;
      requestData["ExceptionType"] = exceptionObj.exception.runtimeType.toString();
    }

    requestData["DeviceID"] = this._deviceId;
    requestData["Stacktrace"] = stacktrace;
    requestData["Severity"] = _getSeverityString(severity);
    requestData["CrashType"] = crashType == CrashType.HANDLED ? "Handled" : "Unhandled";
    requestData["DeviceType"] = "Flutter";
    requestData["ProjectID"] = this._projectId;
    requestData["VersionName"] = this._versionName;

    requestData["ClassFile"] = decodedStack["Class"];
    requestData["LineNo"] = decodedStack["LineNo"];
    requestData["ScreenResolution"] = width.toString() + " x " + height.toString();
    requestData["Locale"] = defaultLocale;
    requestData["OSVersionName"] = osVersion;
    requestData["APIVersion"] = apiVersion;

    if (customProperties.isNotEmpty)
    {
      requestData["CustomProperty"] = jsonEncode(customProperties);
    }
    return requestData;
  }

  HashMap<String, String> _decodeStacktrace(String stack, CrashType crashType) {
    List<String> stackLines = stack.split("\n");

    int stackLineIndex = crashType == CrashType.HANDLED ? 1 : 0;

    String firstLineOfStack = stackLines[stackLineIndex].replaceAll("package:", "");
    int startOfClassLoc = firstLineOfStack.indexOf("(");
    int startOfLineNumber = firstLineOfStack.indexOf(":", startOfClassLoc);
    String classLoc = firstLineOfStack.substring(startOfClassLoc+1, startOfLineNumber);
    String lineNo = firstLineOfStack.substring(startOfLineNumber+1, firstLineOfStack.indexOf(":", startOfLineNumber+1));

    HashMap decodedStack = HashMap<String, String>();
    decodedStack["Class"] = classLoc;
    decodedStack["LineNo"] = lineNo;
    return decodedStack;
  }

  String _getSeverityString(Severity severity)
  {
    switch (severity)
    {
      case Severity.LOW:
        return "Low";
      case Severity.MEDIUM:
        return "Medium";
      case Severity.HIGH:
        return "High";
    }
  }

  void _sendRequest(String endpoint, HashMap<String, dynamic> requestData) async
  {
    String _url = "https://engine.crashcatch.com";
    String requestUrl = _url + "/" + endpoint;

    Map<String, String> requestHeaders;

    if (endpoint == "initialise" || this._sessionId == "") {
      requestHeaders = {
        "Content-Type": "application/x-www-form-urlencoded",
        "authorisation-token": this._apiKey
      };
    }
    else
      {
        requestHeaders = {
          "Content-Type": "application/x-www-form-urlencoded",
          "authorisation-token": this._apiKey,
          "cookie": "SESSIONID=" + this._sessionId
        };
      }

    var response = await http.post(Uri.parse(requestUrl),
      body: requestData,
      headers: requestHeaders
    );

    int statusCode = response.statusCode;
    if (statusCode == 200) {

      if (endpoint == "initialise") {
        Map<String, String> headers = response.headers;

        String cookieString = headers["set-cookie"];
        _parseCookieString(cookieString);

        if (this._sessionId != "") {
          this._initialisationCompleted = true;

          if (CrashCatch._crashQueue.length > 0)
          {
            for (int i = 0; i < CrashCatch._crashQueue.length; i++)
            {
              _sendRequest("crash", CrashCatch._crashQueue[i]);
            }
            CrashCatch._crashQueue.clear();
          }
        }
      }
    }
    else
    {
      this._initialisationCompleted = false;
      this._sessionId = "";
      CrashCatch._crashQueue.add(requestData);
    }

  }

  void _parseCookieString(String cookieString)
  {
    List<String> cookies = cookieString.split(";");
    for (var i = 0; i < cookies.length; i++)
    {
      String cookie = cookies[i];
      if (cookie.startsWith("SESSIONID"))
      {
        List<String> keyValue = cookie.split("=");
        this._sessionId = keyValue[1];
      }
      else if (cookie.startsWith("DOLB"))
      {
        List<String> keyValue = cookie.split("=");
        this._doLb = keyValue[1];
      }
    }
  }


  String generateRandomString() {
    var random = Random.secure();
    var values = List<int>.generate(20, (i) =>  random.nextInt(255));
    return base64UrlEncode(values).toLowerCase();
  }
}


lib/crashcatchlib_flutter.dart
library crashcatchlib_flutter;
import 'dart:async';
import 'dart:collection';
import 'dart:io';
import 'dart:convert';
import 'dart:io' show Platform;
import 'package:flutter/cupertino.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:math';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:web_browser_detect/web_browser_detect.dart';

enum Severity {
  LOW,
  MEDIUM,
  HIGH
}

enum CrashType {
  HANDLED,
  UNHANDLED
}

class CrashCatch
{
  String _projectId = "";
  String _versionName = "";
  String? _deviceId;
  String _apiKey = "";
  String _sessionId = "";
  String _doLb = "";
  bool _initialisationCompleted = false;
  late BuildContext _context;
  static late List<HashMap<String, dynamic>> _crashQueue;

  CrashCatch(BuildContext buildContext) {
    this._context = buildContext;
    CrashCatch._crashQueue = [];
  }

  void _setupUnhandledException() {
    FlutterError.onError = (FlutterErrorDetails details) {
      _sendUnhandledCrash(details);
    };
  }

  void initialiseCrashCatch(String projectId, String apiKey, String version) async
  {
    this._projectId = projectId;
    this._apiKey = apiKey;
    this._versionName = version;

    //Check the app shared preferences to see if the device id has been set, if
    //not generate a random device id string

    final SharedPreferences prefs = await SharedPreferences.getInstance();
    this._deviceId = prefs.getString("crashcatch_device_id") ?? "";

    if (this._deviceId == "")
    {
      this._deviceId = generateRandomString();
      await prefs.setString("crashcatch_device_id", this._deviceId!);
    }

    HashMap requestData = new HashMap<String, dynamic>();
    requestData["ProjectID"] = projectId;
    requestData["DeviceID"] = _deviceId;
    requestData["ProjectVersion"] = version;
     _sendRequest("initialise", requestData as HashMap<String, dynamic>);

    this._setupUnhandledException();

  }

  void reportCrash(Exception exception, Severity severity, {StackTrace? stack, Map<String, dynamic> customProperties = const {} }) async
  {
    String stacktrace = stack == null ? StackTrace.current.toString() : stack
        .toString();

    HashMap<String, String?> requestData = await returnPostData(exception, stacktrace, severity, CrashType.HANDLED, customProperties);

    if (_initialisationCompleted) {
      _sendRequest("crash", requestData);
    }
    else
    {
      CrashCatch._crashQueue.add(requestData);
    }
  }

  void _sendUnhandledCrash(FlutterErrorDetails flutterErrorDetails) async
  {
    HashMap<String, String?> requestData = await returnPostData(flutterErrorDetails, flutterErrorDetails.stack.toString(), Severity.HIGH, CrashType.UNHANDLED, {});
    if (_initialisationCompleted)
    {
      _sendRequest("crash", requestData);
    }
    else
      {
        CrashCatch._crashQueue.add(requestData);
      }
  }

  Future<HashMap<String, String?>> returnPostData(Object exception, String stacktrace, Severity severity, CrashType crashType, Map<String, dynamic> customProperties) async
  {
    HashMap<String, String> decodedStack = _decodeStacktrace(stacktrace, crashType);

    HashMap requestData = new HashMap<String, String?>();

    String osVersion = "";
    String apiVersion = "";

    if (!kIsWeb)
    {
      if (Platform.isAndroid) {
        DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
        AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
        String? version = androidInfo.version.release != null ? androidInfo.version.release : "";
        osVersion = "Android " + version!;
        apiVersion = androidInfo.version.sdkInt.toString();
      }
      else if (Platform.isIOS) {
        osVersion = Platform.operatingSystem;
        apiVersion = Platform.operatingSystemVersion;
      }
      else if (Platform.isWindows)
      {
        osVersion = "Windows";
        apiVersion = Platform.operatingSystemVersion;
      }
    }
    else
    {
      osVersion = "N/A";
      final Browser browserDetails = Browser();

      String browser = browserDetails.browser;
      String? browserVersion = browserDetails.version;

      requestData["Browser"] = browser;
      requestData["BrowserVersion"] = browserVersion;
      String height = MediaQuery.of(_context).size.height.toString();
      String width = MediaQuery.of(_context).size.width.toString();

      requestData["BrowserWidthHeight"] = width + " x " + height;

    }

    int width = MediaQuery
        .of(_context)
        .size
        .width
        .ceil();
    int height = MediaQuery
        .of(_context)
        .size
        .height
        .ceil();
    String defaultLocale = !kIsWeb ? Platform.localeName : "N/A";


    if (exception is Exception)
    {
      Exception exceptionObj = exception;
      requestData["ExceptionType"] = exceptionObj.runtimeType.toString();
      requestData["ExceptionMessage"] = exceptionObj.toString();
    }
    else
    {
      FlutterErrorDetails exceptionObj = exception as FlutterErrorDetails;
      requestData["ExceptionType"] = exceptionObj.exception.runtimeType.toString();
      requestData["ExceptionMessage"] = exceptionObj.exception.toString();
    }

    requestData["DeviceID"] = this._deviceId;
    requestData["Stacktrace"] = stacktrace;
    requestData["Severity"] = _getSeverityString(severity);
    requestData["CrashType"] = crashType == CrashType.HANDLED ? "Handled" : "Unhandled";
    requestData["DeviceType"] = "Flutter";
    requestData["ProjectID"] = this._projectId;
    requestData["VersionName"] = this._versionName;

    requestData["ClassFile"] = decodedStack["Class"];
    requestData["LineNo"] = decodedStack["LineNo"];
    if (!kIsWeb)
    {
      requestData["ScreenResolution"] = width.toString() + " x " + height.toString();
    }
    requestData["Locale"] = defaultLocale;
    requestData["OSName"] = osVersion;
    requestData["OSVersion"] = apiVersion;


    if (customProperties.isNotEmpty)
    {
      requestData["CustomProperty"] = jsonEncode(customProperties);
    }
    return requestData as FutureOr<HashMap<String, String?>> ;
  }

  HashMap<String, String> _decodeStacktrace(String stack, CrashType crashType) {
    List<String> stackLines = stack.split("\n");

    String classLoc = "";
    String lineNo = "";

    if (!kIsWeb) {
      int stackLineIndex = crashType == CrashType.HANDLED ? 1 : 0;

      String firstLineOfStack = stackLines[stackLineIndex].replaceAll("package:", "");
      int startOfClassLoc = firstLineOfStack.indexOf("(");

      int startOfLineNumber = firstLineOfStack.indexOf(":", startOfClassLoc);

      classLoc = firstLineOfStack.substring(startOfClassLoc + 1, startOfLineNumber);

      lineNo = firstLineOfStack.substring(startOfLineNumber + 1,
          firstLineOfStack.indexOf(":", startOfLineNumber + 1));

    }
    else
    {
      int stackLineIndex = 0;
      for (int i = 0; i < stackLines.length; i++) {
          if (stackLines[i].indexOf('dart-sdk') == -1 && stackLines[i].indexOf('crashcatchlib') == -1)
          {
            stackLineIndex = i;
            break;
          }
      }

      String firstLineOfStack = stackLines[stackLineIndex].replaceAll("package:", "");

      int startOfClassLoc = firstLineOfStack.indexOf("/");
      int startOfLineNumber = firstLineOfStack.indexOf(' ', startOfClassLoc) +1;
      int endOfLineNumber = firstLineOfStack.indexOf(":", startOfClassLoc);

      classLoc = firstLineOfStack.substring(startOfClassLoc + 1, firstLineOfStack.indexOf(' ', startOfClassLoc));

      lineNo = firstLineOfStack.substring(startOfLineNumber, endOfLineNumber);

    }

    HashMap decodedStack = HashMap<String, String>();
    decodedStack["Class"] = classLoc;
    decodedStack["LineNo"] = lineNo;
    return decodedStack as HashMap<String, String>;
  }

  String _getSeverityString(Severity severity)
  {
    switch (severity)
    {
      case Severity.LOW:
        return "Low";
      case Severity.MEDIUM:
        return "Medium";
      case Severity.HIGH:
        return "High";
    }
  }

  void _sendRequest(String endpoint, HashMap<String, dynamic> requestData) async
  {
    String _url = "https://engine.crashcatch.com/api";
    String requestUrl = _url + "/" + endpoint;

    Map<String, String> requestHeaders;

    if (endpoint == "initialise" || this._sessionId == "") {

      requestHeaders = {
        "Content-Type": "application/json",
        "authorisation-token": this._apiKey
      };
    }
    else
      {
        String cookieString = "SESSIONID=" + this._sessionId;
        if (this._doLb.length > 0)
        {
          cookieString += "; DO-LB=" + this._doLb;
        }
        requestHeaders = {
          "Content-Type": "application/json",
          "authorisation-token": this._apiKey,
          "session_id": this._sessionId,
          "cookie": cookieString
        };
      }

    var response = await http.post(Uri.parse(requestUrl),
      body: json.encode(requestData),
      headers: requestHeaders
    );

    int statusCode = response.statusCode;
    if (statusCode == 200) {

      if (endpoint == "initialise") {
        Map<String, String> headers = response.headers;

        if (!kIsWeb)
        {
          String cookieString = headers["set-cookie"]!;
          _parseCookieString(cookieString);
        }
        else
        {
            this._sessionId = headers["session_id"]!;
        }

        if (this._sessionId.length != 0) {
          this._initialisationCompleted = true;

          if (CrashCatch._crashQueue.length > 0)
          {
            for (int i = 0; i < CrashCatch._crashQueue.length; i++)
            {
              _sendRequest("crash", CrashCatch._crashQueue[i]);
            }
            CrashCatch._crashQueue.clear();
          }
        }
      }
    }
    else
    {
      this._initialisationCompleted = false;
      this._sessionId = "";
      CrashCatch._crashQueue.add(requestData);
    }

  }

  void _parseCookieString(String cookieString)
  {

    List<String> cookies = cookieString.split(";");

    cookies.addAll(cookieString.split(","));
    for (var i = 0; i < cookies.length; i++)
    {
      String cookie = cookies[i];
      if (cookie.startsWith("SESSIONID"))
      {
        List<String> keyValue = cookie.split("=");
        String value = keyValue[1];
        if (value.contains(";"))
        {
           value = value.substring(0, value.indexOf(";"));
        }
        this._sessionId = value.trim();
      }
      else if (cookie.startsWith("DO-LB"))
      {
        List<String> keyValue = cookie.split("=");
        String value = keyValue[1];
        if (value.contains(";"))
        {
          value = value.substring(0, value.indexOf(";"));
        }
        this._doLb = value;
      }
    }
  }


  String generateRandomString() {
    var random = Random.secure();
    var values = List<int>.generate(20, (i) =>  random.nextInt(255));
    return base64UrlEncode(values).toLowerCase();
  }
}
